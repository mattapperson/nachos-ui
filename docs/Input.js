import React, { Component } from 'react'
import { View } from 'react-native'

import Input from '../src/Input'
import H4 from '../src/typography/H4'

// WEBPACK (InputExample)
class InputExample extends Component {
  state = { value:'' }
  render() {
    const inputStyle = { margin: 15 }
    return (
      <View>
        <H4>Example:</H4>
        <Input
          style={inputStyle}
          placeholder='Your name'
          value={this.state.value}
          onChangeText={value => this.setState({ value })}
        />
        <Input style={inputStyle} disabled value='Disabled input' />
        <Input
          status='warn'
          style={inputStyle}
          placeholder='Your name'
          value={this.state.value}
          onChangeText={value => this.setState({ value })}
        />
        <Input
          status='error'
          style={inputStyle}
          placeholder='Your name'
          value={this.state.value}
          onChangeText={value => this.setState({ value })}
        />
        <Input
          status='valid'
          style={inputStyle}
          placeholder='Your name'
          value={this.state.value}
          onChangeText={value => this.setState({ value })}
        />
        <Input
          icon='ios-beer'
          style={inputStyle}
          placeholder='Your name'
          value={this.state.value}
          onChangeText={value => this.setState({ value })}
        />
      </View>
    )
  }
}
// WEBPACK
const description = `
You can use any of [TextInput](https://facebook.github.io/react-native/docs/textinput.html) props you want.

| Prop | Type | Default | Description |
|------|------|---------|-------------|
| \`value\` | \`string\` | \`''\` | Specifies the value of the Input element|
| \`placeholder\` | \`string\` | \`void\` | Specifies the placeholder value of the Input element|
| \`style\` | \`object, array\` | \`void\` | Style of the Input's View component |
| \`inputStyle\` | \`object, array\` | \`void\` | Style of the TextInput |
| \`icon\` | \`string\` | \`void\` | Add custom [Ionicons](http://ionicframework.com/docs/v2/ionicons/) inside the Input component |
| \`status\` | \`enum\` | \`normal\` | One of: \`normal, valid, error, warn \`|
| \`width\` | \`number\` | \`void\` | Determines a width of the Input compoment |
| \`height\` | \`number\` | \`46\` | Determines a height of the Input compoment |
| \`onChangeText\` | \`func\` | \`void\` | Callback that is called when the input has been changed |
| \`theme\` | \`object\` | \`defaultTheme\` | An object describing a style of the component. See more in Theme section |
`

InputExample.styleguide = {
  ...InputExample.styleguide,
  title: 'Input',
  description,
}

export default InputExample
