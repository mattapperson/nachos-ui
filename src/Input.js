import React from 'react'
import PropTypes from 'prop-types'
import { View, TextInput } from 'react-native'
import Icon from 'react-native-vector-icons/dist/Ionicons'

const defaultTheme = {
  INPUT_HEIGHT: 46,
  INPUT_BACKGROUND: '#fff',
  INPUT_ICON_SIZE: 20,
  INPUT_VALID_ICON: 'md-checkmark',
  INPUT_WARN_ICON: 'md-alert',
  INPUT_ERROR_ICON: 'md-close',
  INPUT_NORMAL_COLOR: '#bdc1cc',
  INPUT_VALID_COLOR: '#66bd2b',
  INPUT_WARN_COLOR: '#ff8c2f',
  INPUT_ERROR_COLOR: '#e03126',
}

const Input = (props) => {
  const { width, theme } = props
  const height = props.height || theme.INPUT_HEIGHT
  const statusStyle = theme[props.status]
  const icon = props.icon || theme.iconStates[props.status]
  const statusColor = theme.colorStates[props.status]

  let IconComponent
  if (icon) {
    IconComponent = (
      <Icon
        name={icon}
        size={theme.INPUT_ICON_SIZE}
        color={statusColor}
        style={[theme.icon]}
      />
    )
  }

  // NOTE: Clone props and then delete Component specific props so we can
  // spread the rest
  let { ...rest } = props
  delete rest.editable
  delete rest.inputStyle
  delete rest.style
  delete rest.disabled
  delete rest.status
  delete rest.icon
  delete rest.height
  delete rest.width

  return (
    <View
      style={[
        theme.base,
        statusStyle,
        props.style,
        { width, height },
        props.disabled ? theme.disabled : {},
      ]}
    >
      <TextInput
        {...rest}
        editable={!props.disabled}
        style={[
          theme.base,
          theme.input,
          props.inputStyle,
          { color: statusColor, width, height },
        ]}
      />
      {IconComponent}
    </View>
  )
}

Input.defaultStyle = (theme = defaultTheme) => {
  return {
    base: { alignSelf: 'stretch', borderWidth: 1 },
    normal: {
      backgroundColor: theme.INPUT_BACKGROUND,
      borderColor: theme.INPUT_NORMAL_COLOR,
      borderStyle: 'solid',
    },
    disabled: { opacity: 0.2 },
    valid: {
      borderColor: theme.INPUT_VALID_COLOR,
      borderStyle: 'solid',
    },
    error: {
      borderColor: theme.INPUT_ERROR_COLOR,
      borderStyle: 'solid',
    },
    warn: {
      borderColor: theme.INPUT_WARN_COLOR,
      borderStyle: 'solid',
    },
    input: {
      borderColor: 'transparent',
      paddingLeft: 12,
      paddingRight: 6,
    },
    icon: {
      top: 12,
      right: 12,
      position: 'absolute',
      backgroundColor: 'transparent',
    },
    iconStates: {
      valid: theme.INPUT_VALID_ICON,
      warn: theme.INPUT_WARN_ICON,
      error: theme.INPUT_ERROR_ICON,
    },
    colorStates: {
      normal: theme.INPUT_NORMAL_COLOR,
      valid: theme.INPUT_VALID_COLOR,
      warn: theme.INPUT_WARN_COLOR,
      error: theme.INPUT_ERROR_COLOR,
    },
  }
}

Input.propTypes = {
  value: PropTypes.string,
  disabled: PropTypes.bool,
  status: PropTypes.oneOf(['normal', 'valid', 'error', 'warn']),
  width: PropTypes.number,
  height: PropTypes.number,
  style: PropTypes.oneOfType([PropTypes.object, PropTypes.array]),
  inputStyle: PropTypes.oneOfType([
    PropTypes.object,
    PropTypes.array,
  ]),
  icon: PropTypes.string,
  theme: PropTypes.object,
}

Input.defaultProps = {
  value: '',
  disabled: false,
  status: 'normal',
}

export default Input
