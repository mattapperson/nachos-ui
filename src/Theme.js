import React, { Component } from "react";
import PropTypes from "prop-types";
import { StyleSheet } from "react-native";
import defaultBranding from "./branding_defaults";

export const ThemeContext = React.createContext({
  theme: {}
});

export class Provider extends React.Component {
  state = {};
  static getDerivedStateFromProps(nextProps, prevState) {
    if (JSON.stringify(nextProps.theme) !== JSON.stringify(prevState)) {
      //return this.createTheme(nextProps.theme);
    }
    return null;
  }

  themeConfigs = {};
  branding = defaultBranding;

  updateThemeConfig = (name, themeConfig) => {
    themeConfig = JSON.parse(
      JSON.stringify(themeConfig).replace(/@([\w_-]+)/gm, (match, key) => {
        if (!this.branding[key]) console.log(key);

        return this.branding[key];
      })
    );

    // if (name === "H1") {
    //   console.log(themeConfig);
    // }
    return (this.themeConfigs[name] = {
      computedStyle: themeConfig.style
        ? StyleSheet.create(themeConfig.style)
        : {},
      props: themeConfig.settings
    });
  };

  render() {
    return (
      <ThemeContext.Provider
        value={{
          theme: this.themeConfigs,
          updateThemeConfig: this.updateThemeConfig
        }}
      >
        {this.props.children}
      </ThemeContext.Provider>
    );
  }
}

export function withTheme(componentName, ThemedComponent) {
  return class extends React.Component {
    constructor() {
      super();
      try {
        ThemeContext._currentValue.updateThemeConfig(
          componentName,
          ThemedComponent.themeConfig
        );
      } catch (e) {
        console.warn(`failed to update theme for ${componentName}`, e);
      }
    }

    isObject(item) {
      return item && typeof item === "object" && !Array.isArray(item);
    }

    mergeDeep(target, ...sources) {
      if (!sources.length) return target;
      const source = sources.shift();

      if (this.isObject(target) && this.isObject(source)) {
        for (const key in source) {
          if (this.isObject(source[key])) {
            if (!target[key]) Object.assign(target, { [key]: {} });
            this.mergeDeep(target[key], source[key]);
          } else {
            Object.assign(target, { [key]: source[key] });
          }
        }
      }

      return this.mergeDeep(target, ...sources);
    }
    render() {
      return (
        <ThemeContext.Consumer>
          {({ theme }) => {
            const computedTheme =
              (theme[componentName] || {}).computedStyle || {};

            const fullTheme = this.props.theme
              ? this.mergeDeep({}, computedTheme, this.props.theme)
              : computedTheme;

            return (
              <ThemedComponent
                {...(theme[componentName] || {}).props || {}}
                {...this.props}
                theme={fullTheme}
              />
            );
          }}
        </ThemeContext.Consumer>
      );
    }
  };
}
