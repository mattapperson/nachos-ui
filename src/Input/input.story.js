import { storiesOf } from "../storybook";
import { withInfo } from "../storybook";
import React from "react";
import Input from "./index";
import { View } from "react-native";

const Example = () => {
  const inputStyle = { margin: 15 };

  return (
    <View style={{ flex: 1, flexDirection: "column" }}>
      <Input style={inputStyle} placeholder="Your name" />
      <Input style={inputStyle} disabled value="Disabled input" />
      <Input status="warn" style={inputStyle} placeholder="Your name" />
      <Input status="error" style={inputStyle} placeholder="Your name" />
      <Input status="valid" style={inputStyle} placeholder="Your name" />
      <Input icon="ios-beer" style={inputStyle} placeholder="Your name" />
    </View>
  );
};

storiesOf("Input", module).add("default", withInfo()(Example));

export default Example;
