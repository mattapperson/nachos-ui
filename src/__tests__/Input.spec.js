import "react-native";
import React from "react";
import renderer from "react-test-renderer";

import Input from "../Input";
import { Provider } from "../Theme";

it("should render correctly", () => {
  const component = renderer.create(
    <Provider>
      <Input value="Your name" />
    </Provider>
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it("should render disabled", () => {
  const component = renderer.create(
    <Provider>
      <Input value="Your name" disabled />
    </Provider>
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
