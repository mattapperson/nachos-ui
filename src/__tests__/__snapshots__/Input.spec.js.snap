exports[`test should render correctly 1`] = `
<View
  style={
    Array [
      Object {
        "alignSelf": "stretch",
        "borderWidth": 1,
      },
      Object {
        "backgroundColor": "#fff",
        "borderColor": "#bdc1cc",
        "borderStyle": "solid",
      },
      undefined,
      Object {
        "height": 46,
        "width": undefined,
      },
      Object {},
    ]
  }>
  <TextInput
    allowFontScaling={true}
    editable={true}
    style={
      Array [
        Object {
          "alignSelf": "stretch",
          "borderWidth": 1,
        },
        Object {
          "borderColor": "transparent",
          "paddingLeft": 12,
          "paddingRight": 6,
        },
        undefined,
        Object {
          "color": "#bdc1cc",
          "height": 46,
          "width": undefined,
        },
      ]
    }
    value="Your name" />
</View>
`;

exports[`test should render disabled 1`] = `
<View
  style={
    Array [
      Object {
        "alignSelf": "stretch",
        "borderWidth": 1,
      },
      Object {
        "backgroundColor": "#fff",
        "borderColor": "#bdc1cc",
        "borderStyle": "solid",
      },
      undefined,
      Object {
        "height": 46,
        "width": undefined,
      },
      Object {
        "opacity": 0.2,
      },
    ]
  }>
  <TextInput
    allowFontScaling={true}
    editable={false}
    style={
      Array [
        Object {
          "alignSelf": "stretch",
          "borderWidth": 1,
        },
        Object {
          "borderColor": "transparent",
          "paddingLeft": 12,
          "paddingRight": 6,
        },
        undefined,
        Object {
          "color": "#bdc1cc",
          "height": 46,
          "width": undefined,
        },
      ]
    }
    value="Your name" />
</View>
`;
